cluster_experiments/experiment_analysis.py

```diff
@@ -680,3 +682,226 @@ def analysis_point_estimate(self, df: pd.DataFrame, verbose: bool = False) -> fl
+        self._data_checks(df=df)
+        return self.analysis_pvalue(df=df)
+
+    def fit_synthetic(self, pre_experiment_df: pd.DataFrame, verbose: bool) -> list:
+        """Returns the weights of each donor"""
+
+        if not any(pre_experiment_df[self.treatment_col] == 1):
+            raise ValueError("No treatment unit found in the data.")
+
+        X = (
+            pre_experiment_df.query(f"{self.treatment_col}==0")
+            .pivot(index=self.cluster_cols, columns=self.time_col)[self.target_col]
+            .T
+        )
+
+        y = (
+            pre_experiment_df.query(f"{self.treatment_col}==1")
+            .pivot(index=self.cluster_cols, columns=self.time_col)[self.target_col]
+            .T.iloc[:, 0]
+        )
+
+        weights = get_w(X, y, verbose)
+
+        return weights
+
+    def fit_predict_synthetic(
+        self,
+        pre_experiment_df: pd.DataFrame,
+        df: pd.DataFrame,
+        treatment_cluster,
+        verbose: bool = False,
+    ) -> pd.DataFrame:
+        """
+        This method adds a column with the synthetic results and filter only the treatment unit.
+
+        First, it calculates the weights of each donor in the control group using the `fit_synthetic` method.
+        It then uses these weights to create a synthetic control group that closely matches the treatment unit before the intervention.
+        The synthetic control group is added to the treatment unit in the dataframe.
+
+        Args:
+            pre_experiment_df (pd.DataFrame): The dataframe containing the data before the intervention.
+            df (pd.DataFrame): The dataframe containing the data after the intervention.
+            treatment_cluster (str): The name of the treatment cluster.
+
+        Returns:
+            pd.DataFrame: The dataframe with the synthetic results added to the treatment cluster.
+        """
+        weights = self.fit_synthetic(
+            pre_experiment_df=pre_experiment_df, verbose=verbose
+        )
+
+        synthetic = (
+            df[self._get_cluster_column(df) != treatment_cluster]
+            .pivot(index=self.time_col, columns=self.cluster_cols)[self.target_col]
+            .values.dot(weights)
+        )
+
+        return df[self._get_cluster_column(df) == treatment_cluster].assign(
+            synthetic=synthetic
+        )  # add synthetic to treatment cluster
+
+    def pvalue_based_on_hypothesis(self, ate: float, avg_effects: dict) -> float:
+        """
+        Returns the p-value of the analysis.
+        1. calculate the average effect after intervention for each unit.
+        2. count how many times the average effect is greater than the real treatment unit
+        3. Divide by the number of units. The result is the p-value using Fisher permutation test.
+        """
+
+        avg_effects = list(avg_effects.values())
+
+        if HypothesisEntries(self.hypothesis) == HypothesisEntries.LESS:
+            return np.mean(avg_effects < ate)
```

Comment:
does this work with a list? I understood avg_effects needs to be an array

Reply:
yep it does. Actually avg_effects comes as a dict, but I had to convert to list to run `np.mean(avg_effects < ate)` . I guess I could have converted it to array as well

Reply:
this is what I'm not following:
```python
>>> import numpy as np
>>> np.mean([1, 2] < 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'list' and 'int'
```

Reply:
it works because the ate is a np.float

<img width="419" alt="image" src="https://github.com/david26694/cluster-experiments/assets/33734080/c000595b-8e72-4164-857e-a70c3c1cf57d">